File: src/main/java/org/programmerplanet/sshtunnel/ui/SessionConnectionMonitorFactory.java
package org.programmerplanet.sshtunnel.ui;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.programmerplanet.sshtunnel.model.Session;

/**
 * 
 * @author <a href="agungm@outlook.com">Mulya Agung</a>
 */

public class SessionConnectionMonitorFactory {

	private static final Log log = LogFactory.getLog(SessionConnectionMonitorFactory.class);

	private static final int DEF_MONITOR_INTERVAL = 10000;
	private static final SessionConnectionMonitorFactory INSTANCE = new SessionConnectionMonitorFactory();

	private final Object lock = new Object();
	private Thread thread;
	private final SessionConnectionMonitor sessionConnectionMonitor;

	public SessionConnectionMonitorFactory() {
		this(DEF_MONITOR_INTERVAL);
	}

	public SessionConnectionMonitorFactory(int monitorInterval) {
		this.sessionConnectionMonitor = new SessionConnectionMonitor(monitorInterval);
	}

	public void startMonitor() {
		synchronized (lock) {
			if (thread == null) {
				thread = new Thread(this.sessionConnectionMonitor);
				thread.start();
			}
		}
	}

	public void stopMonitor() {
		synchronized (lock) {
			if (thread != null) {
				this.sessionConnectionMonitor.setThreadStopped(true);
				thread = null;
				
				if (log.isWarnEnabled()) {
					log.warn("Connection monitor is stopped.");
				}
			}
		}
	}

	public void addSession(String name, Session session) {
		this.sessionConnectionMonitor.addSession(name, session);
	}

	public void removeSession(String name) {
		this.sessionConnectionMonitor.removeSession(name);
	}

	public void setSshTunnelComposite(SshTunnelComposite sshTunnelComposite) {
		this.sessionConnectionMonitor.setSshTunnelComposite(sshTunnelComposite);
	}

	public static SessionConnectionMonitorFactory getInstance() {
		return INSTANCE;
	}
}

File: src/main/kotlin/org/programmerplanet/sshtunnel/ui/SessionConnectionMonitorFactory.kt
package org.programmerplanet.sshtunnel.ui

import io.github.oshai.kotlinlogging.KotlinLogging
import org.programmerplanet.sshtunnel.model.Session

private val logger = KotlinLogging.logger {}

/**
 *
 * @author [Mulya Agung](agungm@outlook.com)
 */
object SessionConnectionMonitorFactory {
    private const val DEF_MONITOR_INTERVAL = 10000

    private val lock = Any()
    private var thread: Thread? = null
    private val sessionConnectionMonitor = SessionConnectionMonitor(DEF_MONITOR_INTERVAL)

    fun startMonitor() {
        synchronized(lock) {
            if (thread == null) {
                thread = Thread(sessionConnectionMonitor).apply {
                    start()
                }
            }
        }
    }

    fun stopMonitor() {
        synchronized(lock) {
            thread?.let {
                sessionConnectionMonitor.setThreadStopped(true)
                thread = null
                if (logger.isWarnEnabled()) {
                    logger.warn {"Connection monitor is stopped."}
                }
            }
        }
    }

    fun addSession(name: String, session: Session) {
        sessionConnectionMonitor.addSession(name, session)
    }

    fun removeSession(name: String) {
        sessionConnectionMonitor.removeSession(name)
    }

    fun setSshTunnelComposite(applicationComposite: ApplicationComposite) {
        sessionConnectionMonitor.setSshTunnelComposite(applicationComposite)
    }
}
